xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,256 @@
+ * USA
+ */
+
+/*
+ListNET is adapted from the paper:
+Cao, Zhe, et al. "Learning to rank: from pairwise approach to listwise approach."
+Proceedings of the 24th international conference on Machine learning. ACM, 2007.
+*/
+
+#include "xapian-letor/ranker.h"
+#include "debuglog.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+using namespace Xapian;
+
+/// A vector holding ground truth and prediction score probability distribution vectors.
+typedef vector <vector <double> > prob_distrib_vector;
+
+ListNETRanker::~ListNETRanker() {
+    LOGCALL_DTOR(API, "ListNETRanker");
+}
+
+static double
+calculateInnerProduct(vector<double> parameters, vector<double> fvals) {
+    LOGCALL_STATIC_VOID(API, "calculateInnerProduct", parameters | fvals);
+
+    double inner_product = 0.0;
+
+    for (int i = 0; i < int(fvals.size()); ++i){
+	inner_product += parameters[i] * fvals[i];
+    }
+
+    return inner_product;
+}
+
+// From Theorem (8) in Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static prob_distrib_vector
+initializeProbability(vector<FeatureVector> feature_vectors, vector<double> & new_parameters) {
+    LOGCALL_STATIC_VOID(API, "initializeProbability", feature_vectors | new_parameters);
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+	expsum_y += exp(feature_vectors[i].get_label());
+	expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+	prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+	prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+    return prob;
+}
+
+// Equation (6) in paper Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, prob_distrib_vector prob) {
+    LOGCALL_STATIC_VOID(API, "calculateGradient", feature_vectors | prob_distrib_vector);
+
+    vector<double> gradient(feature_vectors[0].get_fcount(),0);
+    int list_length = feature_vectors.size();
+
+    // Hold ground truth probability distribution
+    vector<double> prob_y(prob[0]);
+    // Hold prediction score probability distribution
+    vector<double> prob_z(prob[1]);
+
+    for (int i = 0; i < list_length; i++) {
+	vector<double> fvals = feature_vectors[i].get_fvals();
+
+	for (int k = 0; k < int(fvals.size()); k++){
```

Comment:
It would be cleaner to make the `k` the same type as `fvals.size()` - i.e. `fvals::size_type`.  Then you don't need to cast things to `int` to avoid warnings, and you don't even need to consider what happens if the cast truncates the value.

Similarly for `i` and `list_length`.

Also, missing space before `{`.